optimization.ts

```diff
@@ -0,0 +1,433 @@
+import { BinOp, Parameter, Type, UniOp} from "./ast";
+import { Stmt, Expr, Value, VarInit, BasicBlock, Program, FunDef, Class } from "./ir";
+
+import { isTagBoolean, isTagNone, isTagId, isTagBigInt, isTagEqual, checkValueEquality, checkCompileValEquality, checkStmtEquality } from "./optimization_utils"; 
+
+type Env = {
+    vars: Map<string, compileVal>;
+}
+
+export type compileVal = {
+    tag: "nac"|"val"|"undef", value?: Value<any>;
+}
+
+export function optimizeValue(val: Value<any>, env: Env): Value<any>{
+    if (val.tag !== "id"){
+        return val;
+    }
+    if (env.vars.has(val.name)){
+        if (["nac", "undef"].includes(env.vars.get(val.name).tag))
+            return val;
+        val = env.vars.get(val.name).value;
+    }
+    return val;
+}
+
+export function evaluateBinOp(op: BinOp, leftVal: Value<any>, rightVal: Value<any>): Value<any>{
+    if([BinOp.Plus, BinOp.Minus,BinOp.IDiv,BinOp.Mul, BinOp.Gt, BinOp.Lt, BinOp.Gte, BinOp.Lte, BinOp.Mod].includes(op)){
+        if(!isTagBigInt(leftVal) || !isTagBigInt(rightVal))
+            throw new Error("Compiler Error");
+        
+        switch(op){
+            case BinOp.Plus: return {tag: "num", value: leftVal.value + rightVal.value};
```

Comment:
bigints to the rescue! I guess that saves from checking overflow and a bunch of other annoying cases.

Reply:
True!
At the same time, towards the end if time permits, we could modify IR generation to not typecast every number as bigint depending on how tedious the changes the can get and also depending on how efficiently bignums have been implemented (if at the implementation, it behaves as an i32/i64 when it fits into the range, then I'm guessing there should be only the additional overhead at compile time).